CRM/Core/EntityTokens.php

```diff
@@ -284,4 +305,101 @@ public function alterActionScheduleQuery(MailingQueryEvent $e): void {
     }
   }
 
+  /**
+   * Get tokens supporting the syntax we are migrating to.
+   *
+   * In general these are tokens that were not previously supported
+   * so we can add them in the preferred way or that we have
+   * undertaken some, as yet to be written, db update.
+   *
+   * See https://lab.civicrm.org/dev/core/-/issues/2650
+   *
+   * @return string[]
+   * @throws \API_Exception
+   */
+  public function getBasicTokens(): array {
+    $return = [];
+    foreach ($this->getExposedFields() as $fieldName) {
+      $return[$fieldName] = $this->getFieldMetadata()[$fieldName]['title'];
+    }
+    return $return;
+  }
+
+  /**
+   * Get entity fields that should be exposed as tokens.
+   *
+   * @return string[]
+   *
+   */
+  public function getExposedFields(): array {
+    $return = [];
+    foreach ($this->getFieldMetadata() as $field) {
+      if (!in_array($field['name'], $this->getSkippedFields(), TRUE)) {
+        $return[] = $field['name'];
+      }
+    }
+    return $return;
+  }
+
+  /**
+   * Get entity fields that should not be exposed as tokens.
+   *
+   * @return string[]
+   */
+  public function getSkippedFields(): array {
+    $fields = ['contact_id'];
+    if (!CRM_Campaign_BAO_Campaign::isCampaignEnable()) {
+      $fields[] = 'campaign_id';
+    }
+    return $fields;
+  }
+
+  /**
+   * @return string
+   */
+  protected function getEntityName(): string {
+    return CRM_Core_DAO_AllCoreTables::convertEntityNameToLower($this->getApiEntityName());
+  }
+
+  public function getEntityIDField() {
+    return $this->getEntityName() . 'Id';
+  }
+
+  public function prefetch(\Civi\Token\Event\TokenValueEvent $e): ?array {
+    $entityIDs = $e->getTokenProcessor()->getContextValues($this->getEntityIDField());
+    if (empty($entityIDs)) {
+      return [];
+    }
+    $select = $this->getPrefetchFields($e);
+    $result = (array) civicrm_api4($this->getApiEntityName(), 'get', [
+      'checkPermissions' => FALSE,
+      // Note custom fields are not yet added - I need to
+      // re-do the unit tests to support custom fields first.
+      'select' => $select,
+      'where' => [['id', 'IN', $entityIDs]],
+    ], 'id');
+    return $result;
+  }
+
+  public function getCurrencyFieldName() {
+    return [];
+  }
+
+  /**
+   * Get the currency to use for formatting money.
+   * @param $row
+   *
+   * @return string
+   */
+  public function getCurrency($row): string {
```

Comment:
This asserts a strong opinion that currency is one-per-row. While the `civicrm_contribution` record meets that expectation, I don't see why it's a good thing to bake that opinion into the base-class of token source.

Reply:
Yeah - it's also overrideable by class. I can't offhand think of an exception to there being zero or one per row as this expects - but if there were more than one it could be added